gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/stream/LiKafkaByteArrayMsgSchemaChangeInjector.java

```diff
@@ -0,0 +1,131 @@
+  public ControlMessageInjector<S, byte[]> init(WorkUnitState workUnitState) {
+    Preconditions.checkArgument(workUnitState.contains(KafkaSchemaRegistryConfigurationKeys.KAFKA_SCHEMA_REGISTRY_CLASS),
+        "Missing required property " + KafkaSchemaRegistryConfigurationKeys.KAFKA_SCHEMA_REGISTRY_CLASS);
+
+    this.schemaRegistry = KafkaSchemaRegistryFactory.getSchemaRegistry(workUnitState.getProperties());
+    this.schemaIdCache = CacheBuilder.newBuilder()
+        .expireAfterAccess(1, TimeUnit.HOURS)
+        .build();
+    this.topicName = workUnitState.getProp(KafkaSource.TOPIC_NAME);
+
+    return this;
+  }
+
+  @Override
+  public void setInputGlobalMetadata(GlobalMetadata<S> inputGlobalMetadata, WorkUnitState workUnitState) {
+    this.globalMetadata = inputGlobalMetadata;
+  }
+
+  /**
+   * Inject a {@link MetadataUpdateControlMessage} if the latest schema has changed.
+   * @param inputRecordEnvelope input record envelope
+   * @param workUnitState work unit state
+   * @return the injected messages
+   */
+  @Override
+  public Iterable<ControlMessage<byte[]>> injectControlMessagesBefore(RecordEnvelope<byte[]> inputRecordEnvelope,
+                                                                      WorkUnitState workUnitState) {
+    byte[] data = inputRecordEnvelope.getRecord();
+
+    // magic byte should be there for messages from LiKafka
+    if (inputRecordEnvelope.getRecord()[0] != LiAvroSerDeHelper.MAGIC_BYTE) {
+      throw new RuntimeException(String.format("Unknown magic byte for topic: %s ", topicName));
+    }
+
+    // New schema id so check if the latest schema in the registry has changed.
+    // Only check for the latest schema when a new schema id is seen since the call to get the latest schema is not
+    // cacheable and is expensive.
+    if (this.schemaIdCache.getIfPresent(ByteBuffer.wrap(data, 1, MD5Digest.MD5_BYTES_LENGTH)) == null) {
+      try {
+        S latestSchema = (S) this.schemaRegistry.getLatestSchema(topicName);
+
+        // copy to a new byte buffer to avoid holding a reference to the whole record
+        this.schemaIdCache.put(ByteBuffer.wrap(Arrays.copyOfRange(data, 1, MD5Digest.MD5_BYTES_LENGTH + 1)), "");
+
+        // latest schema changed, so inject a metadata update control message
+        if (!latestSchema.equals(this.globalMetadata.getSchema())) {
+          // update the metadata in this injector since the control message is only applied downstream
+          this.globalMetadata = GlobalMetadata.<S, S>builderWithInput(this.globalMetadata,
+              Optional.of(latestSchema)).build();
+
+          // inject a metadata update control message followed by a flush control message since the avro data needs to
+          // be flushed and a new file opened to write data in the new schema
+          return Lists.newArrayList(new MetadataUpdateControlMessage<S, byte[]>(this.globalMetadata),
+              FlushControlMessage.<byte[]>builder().flushReason("Metadata change")
```

Comment:
Why do you have to send a flush? Can't the writer trigger the flush itself when it sees a metadata change?

Reply:
Yes, that is another way of doing it. The CloseOnWriteWriterWrapper can handle the message. I thought about both ways and was originally going to do it that way, but then decided to use the existing flush handlers since that is more generic than having the logic branch in the control message handler.

Reply:
The Avro case case requires the file to close since the file format does not support evolution, but there may be other formats that support evolution, so I'll move the logic to the writer to allow flexibility for this case.